ppoprf/src/ggm.rs

```diff
@@ -158,63 +152,54 @@ impl PPRF for GGM {
 
-    fn puncture(&mut self, input: &[u8]) {
+    fn puncture(&mut self, input: &[u8]) -> Result<(), PPRFError> {
         if input.len() != self.inp_len {
-            panic!(
-                "Input length ({}) does not match input param ({})",
-                input.len(),
-                self.inp_len
-            );
+            return Err(PPRFError::BadInputLength {
+                actual: input.len(),
+                expected: self.inp_len,
+            });
         }
         let bv = bvcast_u8_to_usize(&BitVec::<Lsb0, _>::from_slice(input).unwrap());
-        if let Ok(pfx) = self.key.find_prefix(&bv) {
-            let pfx_len = pfx.0.len();
-
-            // If the prfix is smaller than the current input, then we
-            // need to recompute some parts of the tree. Otherwise we
-            // just remove the prefix entirely.
-            let mut new_pfxs: Vec<(Prefix, Vec<u8>)> = Vec::new();
-            if pfx_len != bv.len() {
-                let mut iter_bv = bv.clone();
-                for i in (0..bv.len()).rev() {
-                    if let Some((last, rest)) = iter_bv.clone().split_last() {
-                        let mut cbv = iter_bv.clone();
-                        cbv.set(i, !*last);
-                        let mut out = vec![0u8; 32];
-                        let (_, split) = cbv.split_at(pfx_len);
-                        self.bit_eval(&split.to_bitvec(), &pfx.1, &mut out);
-                        new_pfxs.push((Prefix::new(cbv), out));
-                        if rest.len() == pfx_len {
-                            // we don't want to recompute any further
-                            break;
-                        }
-                        iter_bv = rest.to_bitvec();
-                    } else {
-                        panic!("Unexpected end of input");
+        let pfx = self.key.find_prefix(&bv)?;
+        let pfx_len = pfx.0.len();
+
+        // If the prfix is smaller than the current input, then we
+        // need to recompute some parts of the tree. Otherwise we
+        // just remove the prefix entirely.
+        let mut new_pfxs: Vec<(Prefix, Vec<u8>)> = Vec::new();
+        if pfx_len != bv.len() {
+            let mut iter_bv = bv.clone();
+            for i in (0..bv.len()).rev() {
+                if let Some((last, rest)) = iter_bv.clone().split_last() {
+                    let mut cbv = iter_bv.clone();
+                    cbv.set(i, !*last);
+                    let mut out = vec![0u8; 32];
+                    let (_, split) = cbv.split_at(pfx_len);
+                    self.bit_eval(&split.to_bitvec(), &pfx.1, &mut out);
+                    new_pfxs.push((Prefix::new(cbv), out));
+                    if rest.len() == pfx_len {
+                        // we don't want to recompute any further
+                        break;
                     }
+                    iter_bv = rest.to_bitvec();
+                } else {
+                    panic!("Unexpected end of input");
```

Comment:
Maybe it would make sense to create an error for this panic as well?

Reply:
Yes, I had this on my to-do list. Will work on adding that